workflows/host-genome-generation/host_genome_generation.wdl

```diff
@@ -1,150 +1,341 @@
+    docker_image_id,
+  }
+
+  call star_generate {
+    input:
+    fasta = concatenate_and_unzip_fastas.fasta,
+    transcripts_gtf_gz,
+    genome_name,
+    docker_image_id,
+  }
+
+  output {
+    File bowtie2_index_tar = bowtie2_build.index_tar
+    File hisat2_index_tar = hisat2_build.index_tar
+    File kallisto_idx = kallisto_index.idx
+    File minimap2_dna_mmi = minimap2_index_dna.index_mmi
+    File minimap2_rna_mmi = minimap2_index_rna.index_mmi
+    File star_genome_tar = star_generate.star_genome_tar
+
+    # also output the input files, to facilitate archival/provenance
+    File original_genome_fasta_gz = genome_fasta.gz
+    File? original_transcripts_gtf_gz = transcripts_gtf_gz
+    Array[File] original_transcripts_fasta_gz = transcripts_fasta_gz
+    File original_ERCC_fasta_gz = ERCC_fasta_gz
+    Array[File] original_other_fasta_gz = other_fasta_gz
+  }
+}
+
+task ensure_gz {
+  input {
+    File maybe_gz
+    String docker_image_id
+  }
+
+  String name = basename(maybe_gz)
+
+  command <<<
+    set -euxo pipefail
+    mkdir ans
+    if gzip -t '~{maybe_gz}'; then
+      cp '~{maybe_gz}' ans/
+    else
+      pigz -c -p 4 '~{maybe_gz}' > 'ans/~{name}.gz'
+    fi
+  >>>
+
+  output {
+      File gz = glob("ans/*")[0]
+  }
+
+  runtime {
+      docker: docker_image_id
+      cpu: 4
+      memory: "4GiB"
+  }
+}
+
+task concatenate_and_unzip_fastas {
+  input {
+    Array[File] fasta_gz
+    String docker_image_id
+  }
+
+  command <<<
+    pigz -dc ~{sep(' ',fasta_gz)} > "all.fasta"
+  >>>
+
+  output {
+      File fasta = "all.fasta"
+  }
+
+  runtime {
+      docker: docker_image_id
+      cpu: 4
+      memory: "4GiB"
+  }
+}
+
+task bowtie2_build {
+  input {
+    File fasta
+    String genome_name
+    Int seed = 42
+
+    Int cpu = 16
+    String docker_image_id
+  }
+
+  command <<<
+    set -euxo pipefail
+    TMPDIR=${TMPDIR:-/tmp}
+
+    mkdir -p "$TMPDIR"'/bt2/~{genome_name}'
+    >&2 bowtie2-build --seed ~{seed} --threads ~{cpu} "~{fasta}" "$TMPDIR"'/bt2/~{genome_name}/~{genome_name}'
+    >&2 ls -lR "$TMPDIR/bt2"
+    env -C "$TMPDIR/bt2" tar c . > '~{genome_name}.bowtie2.tar'
+  >>>
+
+  output {
+      File index_tar = "~{genome_name}.bowtie2.tar"
+  }
+
+  runtime {
+      docker: docker_image_id
+      cpu: cpu
+      memory: "~{cpu*2}GiB"
+  }
+}
+
+task hisat2_build {
+  input {
+    File fasta
+    File? transcripts_gtf_gz
+    String genome_name
+
+    Int cpu = 32
+    String docker_image_id
+  }
+
+  command <<<
+    set -euxo pipefail
+    TMPDIR=${TMPDIR:-/tmp}
+
+    mkdir -p "$TMPDIR"'/hisat2/~{genome_name}'
+    if [[ -n '~{transcripts_gtf_gz}' ]]; then
+      # convert GTF per http://daehwankimlab.github.io/hisat2/howto/
+      /hisat2/hisat2_extract_splice_sites.py <(pigz -dc '~{transcripts_gtf_gz}') > "$TMPDIR/genome.ss" & pid=$!
+      /hisat2/hisat2_extract_exons.py <(pigz -dc '~{transcripts_gtf_gz}') > "$TMPDIR/genome.exon"
+      wait $pid
+      >&2 /hisat2/hisat2-build -p 16 \
```

Comment:
@mlin @morsecodist  I'm having trouble with this build when the only `transcripts_gtf_gz` is the ERCC.gtf. The hisat2-build command fails with `Error: Encountered exception: 'Nongraph exception'`

The ERCC.gtf is needed for STAR to properly run.

Reply:
Can we say if the file `TMPDIR/genome.ss` don't bother running with `--exon "$TMPDIR/genome.exon" --ss "$TMPDIR/genome.ss"`?